// bundlemon@3.1.0/lib/main/utils/ci/providers/github.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../utils");
const provider = {
    isItMe: !!(0, utils_1.getEnvVar)('GITHUB_ACTION'),
    getVars: () => {
        const raw = (0, utils_1.envVarsListToObject)([
            'GITHUB_REPOSITORY',
            'GITHUB_EVENT_NAME',
            'GITHUB_REF',
            'GITHUB_HEAD_REF',
            'GITHUB_SHA',
            'GITHUB_BASE_REF',
            'GITHUB_RUN_ID',
        ]);
        const fullRepoName = raw.GITHUB_REPOSITORY;
        const [owner, repo] = fullRepoName?.split('/') ?? [undefined, undefined];
        const event = raw.GITHUB_EVENT_NAME;
        const isPr = event === 'pull_request';
        const ref = raw.GITHUB_REF?.split('/');
        return {
            raw,
            ci: true,
            provider: 'github',
            owner,
            repo,
            branch: isPr ? raw.GITHUB_HEAD_REF : ref?.slice(2).join('/'),
            commitSha: raw.GITHUB_SHA,
            targetBranch: raw.GITHUB_BASE_REF,
            prNumber: isPr ? ref?.[2] : undefined,
            buildId: raw.GITHUB_RUN_ID,
        };
    },
};
exports.default = provider;
