// bundlemon@3.1.0/lib/main/utils/ci/providers/travis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../utils");
// https://docs.travis-ci.com/user/environment-variables#default-environment-variables
const provider = {
    isItMe: (0, utils_1.getEnvVar)('TRAVIS') === 'true',
    getVars: () => {
        const raw = (0, utils_1.envVarsListToObject)([
            'TRAVIS_REPO_SLUG',
            'TRAVIS_EVENT_TYPE',
            'TRAVIS_PULL_REQUEST',
            'TRAVIS_BRANCH',
            'TRAVIS_PULL_REQUEST_BRANCH',
            'TRAVIS_COMMIT',
            'TRAVIS_COMMIT_MESSAGE',
        ]);
        const fullRepoName = raw.TRAVIS_REPO_SLUG;
        const [owner, repo] = fullRepoName?.split('/') ?? [undefined, undefined];
        const isPushEvent = raw.TRAVIS_EVENT_TYPE === 'push';
        const prNumber = raw.TRAVIS_PULL_REQUEST;
        return {
            raw,
            ci: true,
            provider: 'travis',
            owner,
            repo,
            branch: isPushEvent ? raw.TRAVIS_BRANCH : raw.TRAVIS_PULL_REQUEST_BRANCH,
            commitSha: raw.TRAVIS_COMMIT,
            targetBranch: isPushEvent ? undefined : raw.TRAVIS_BRANCH,
            prNumber: prNumber === 'false' ? undefined : prNumber, // "false" if it’s not a pull request, set as undefined
            commitMsg: raw.TRAVIS_COMMIT_MESSAGE,
        };
    },
};
exports.default = provider;
