// bundlemon@3.1.0/lib/common/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_PATH_LABELS = exports.CreateCommitRecordAuthType = exports.version = exports.serviceUrl = exports.EnvVar = void 0;
var EnvVar;
(function (EnvVar) {
    EnvVar["remoteFlag"] = "BUNDLEMON_REMOTE";
    EnvVar["projectId"] = "BUNDLEMON_PROJECT_ID";
    EnvVar["projectApiKey"] = "BUNDLEMON_PROJECT_APIKEY";
    EnvVar["serviceURL"] = "BUNDLEMON_SERVICE_URL";
    EnvVar["subProject"] = "BUNDLEMON_SUB_PROJECT";
})(EnvVar || (exports.EnvVar = EnvVar = {}));
exports.serviceUrl = process.env[EnvVar.serviceURL] || 'https://api.bundlemon.dev';
// eslint-disable-next-line @typescript-eslint/no-var-requires
const packageJSON = require('../../package.json');
exports.version = packageJSON.version;
var CreateCommitRecordAuthType;
(function (CreateCommitRecordAuthType) {
    CreateCommitRecordAuthType["ProjectApiKey"] = "PROJECT_API_KEY";
    CreateCommitRecordAuthType["GithubActions"] = "GITHUB_ACTIONS";
})(CreateCommitRecordAuthType || (exports.CreateCommitRecordAuthType = CreateCommitRecordAuthType = {}));
exports.DEFAULT_PATH_LABELS = {
    hash: '[a-zA-Z0-9]+',
};
